src/modules/lost/functions.c

```diff
@@ -580,116 +930,123 @@ int lost_function(struct sip_msg *_m, char *_con, char *_uri, char *_name,
 			/* ! functions.c => http_client_query => query_params.oneline = 0; */
-			curlres =
-					httpapi.http_client_query(_m, url.s, &pidfhdr, NULL, NULL);
-			/* free memory */
-			url.s = NULL;
-			url.len = 0;
+			if(held_post_req == 0) {
+				curl = httpapi.http_client_query_c(
+						_m, url.s, &ret, NULL, mtheld, ACCEPT_HDR);
+			} else {
+				len = 0;
+				heldreq = lost_held_post_request(&len, 0, NULL);
+				if(len > 0) {
+
+					LM_DBG("POST request: [%.*s]\n", len, heldreq);
+
+					curl = httpapi.http_client_query_c(
+							_m, url.s, &ret, heldreq, mtheld, ACCEPT_HDR);
+					pkg_free(heldreq); /* clean up */
+				} else {
+					LM_ERR("could not create POST request\n");
+					goto err;
+				}
+			}
 			/* only HTTP 2xx responses are accepted */
-			if(curlres >= 300 || curlres < 100) {
-				LM_ERR("http GET failed with error: %d\n", curlres);
-				/* free memory */
-				lost_delete_geoheader_list(geolist);
-				lost_free_string(&pidfhdr);
-				lost_free_string(&geohdr);
+			if(curl >= 300 || curl < 100) {
+				if(held_post_req == 0) {
+					LM_ERR("GET [%.*s] failed with error: %d\n", url.len, url.s,
+							curl);
+				} else {
+					LM_ERR("POST [%.*s] failed with error: %d\n", url.len,
+							url.s, curl);
+				}
+				/* clean up */
+				lost_free_string(&ret);
 				goto err;
 			}
-
-			pidf.s = pidfhdr.s;
-			pidf.len = pidfhdr.len;
-
-			if(pidf.len > 0) {
+			url.s = NULL;
+			url.len = 0;
+			pidf.s = ret.s;
+			pidf.len = ret.len;
+			if(pidf.s != NULL && pidf.len > 0) {
 
 				LM_DBG("LbR pidf-lo: [%.*s]\n", pidf.len, pidf.s);
 
-				/* parse the pidf-lo */
-				loc = lost_parse_pidf(pidf, urn);
-				/* free memory */
-				pidf.s = NULL;
-				pidf.len = 0;
 			} else {
 				LM_WARN("dereferencing location failed\n");
 			}
 		}
-		/* free memory */
-		lost_delete_geoheader_list(geolist);
-		lost_free_string(&geohdr);
-		lost_free_string(&pidfhdr);
+		/* clean up */
+		lost_free_geoheader_list(&geolist);
+		lost_free_string(&ret);
 	}
-
-	if(loc == NULL) {
+	if(pidf.s == NULL && pidf.len == 0) {
 		LM_ERR("location object not found\n");
 		goto err;
 	}
-
-	/* check if connection exits */
-	if(httpapi.http_connection_exists(&con) == 0) {
-		LM_ERR("connection: [%.*s] does not exist\n", con.len, con.s);
+	/* parse the pidf and get loc object */
+	loc = lost_parse_pidf(pidf, urn);
+	if(loc == NULL) {
+		LM_ERR("parsing pidf failed\n");
 		goto err;
 	}
 	/* assemble findService request */
 	req.s = lost_find_service_request(loc, &req.len);
-	/* free memory */
-	lost_free_loc(loc);
-	loc = NULL;
-
-	if(req.len == 0) {
+	lost_free_loc(&loc); /* clean up */
+	if(req.s == NULL && req.len == 0) {
 		LM_ERR("lost request failed\n");
 		goto err;
 	}
 
 	LM_DBG("findService request: [%.*s]\n", req.len, req.s);
 
 	/* send findService request to mapping server - HTTP POST */
-	curlres = httpapi.http_connect(_m, &con, NULL, &ret, mtlost, &req);
+	if(naptr) {
+		/* copy url */
+		len = 0;
+		urlrep = lost_copy_string(url, &len);
```

Comment:
same

Reply:
curl requires a zero terminated string as url, which is not added by the http_client API query function `http_client_query_c`